Feat: 틈 관련 API의 날짜 걸침 일정 및 자정 처리

yangjiae12 opened the pull request:
### 👀 관련 이슈

<!-- 관련 이슈를 적어주세요 -->
#114 

### ✨ 작업한 내용

<!-- 작업한 내용을 적어주세요 -->
- 종료 시간이 `00:00`일 경우 `24:00`으로 반환, DB 상으로는 다음 날 `00:00`
- `parseTimeForCompare`, `parseTimeForSort` 함수 분리로 정렬/병합 용도 분리
- 수면 시간이 `23:00 ~ 07:00`처럼 하루를 넘는 경우, `24:00`까지 정확히 반영되도록 수정
- `fromSchedule()` → `sliceScheduleToDate()`로 대체
  - 하루를 넘는 일정인 경우 해당 날짜에 해당하는 구간만 슬라이스하여 TimeSlot 반환
- `Schedule` 조회 로직 수정
  - 기존 `startTime.date == date`에서 `start < 내일 && end ≥ 오늘` 조건으로 조회 범위 확장

### 🌀 PR Point

<!-- 코드리뷰가 필요한 부분이 있다면 적어주세요 -->
- 하루를 넘는 일정이 제대로 잘려 반영되는지
- 자정 처리 관련해서 문제가 없는지

### 🍰 참고사항

<!-- 참고할 사항이 있다면 적어주세요 -->
- 지난 회의 때 얘기 나왔던 것 중에 자정 처리만 하면 된다고 생각하고 이슈를 팠다가 날짜 걸침 일정까지 한꺼번에 고려하게 되면서 공통 시간대 조회 API가 좀 많이 복잡해졌습니다...
- 프론트 측에서 종료 시각의 자정은 입력 시 `00:00`, 반환 시 `24:00`로 달라고 하여서 그거 반영해서 구현했습니다!
- `23:00 ~ 01:00` 일정은 → `"23:00 ~ 24:00"`, `"00:00 ~ 01:00"`으로 각 날짜에 나눠 반영됨
- `sliceScheduleToDate()`는 날짜 단위로 자르고, 정렬/병합은 시간 단위로 수행됨

### 📷 테스트

#### 틈 요청 시 종료 시각에 00:00 입력
```json
{
  "isSuccess": true,
  "code": "TEUM2000",
  "message": "틈 요청이 성공적으로 생성되었습니다.",
  "result": 1
}
```
#### 틈 요청 조회 시 종료 시각에 24:00 반환
```json
{
  "isSuccess": true,
  "code": "TEUM2002",
  "message": "받은 틈 요청 목록이 조회되었습니다.",
  "result": {
    "content": [
      {
        "responseId": 1,
        "requestId": 1,
        "title": "string",
        "description": "string",
        "graphicId": 1,
        "senderUser": {
          "userId": 2,
          "nickname": "유저2",
          "profileImageUrl": "https://teumteum..."
        },
        "receiverCount": 1,
        "date": "2025-08-09",
        "timeSlot": {
          "start": "23:00",
          "end": "24:00"  // 24시 반환
        },
        "read": false
      }
    ], 
    // 중략...
}
```
#### 약속된 틈 조회 시 종료 시각에 24:00 반환
```json
{
  "isSuccess": true,
  "code": "TEUM2009",
  "message": "약속된 틈 상세 정보가 조회되었습니다.",
  "result": {
    "teumId": 1,
    "title": "string",
    "date": "2025-08-09",
    "startTime": "23:00",
    "endTime": "24:00",  // 24시 반환
    "status": "ACTIVE",
    "participants": [
      {
        "userId": 1,
        "nickname": null,
        "profileImageUrl": "https://teumteum..."
      },
      {
        "userId": 2,
        "nickname": "유저2",
        "profileImageUrl": "https://teumteum..."
      }
    ]
  }
}
```

#### 공통 시간대 조회 (날짜 걸침 일정 고려)
```json
{
  "isSuccess": true,
  "code": "TEUM2011",
  "message": "공통 가능한 시간대가 조회되었습니다.",
  "result": {
    "date": "2025-08-01",
    "availableTime": [
      {
        "start": "00:00",
        "end": "02:10"
      },
      {
        "start": "08:10",
        "end": "23:00"
      },
      {
        "start": "23:30",
        "end": "23:40"
      }
    ]
  }
}
```
```json
{
  "isSuccess": true,
  "code": "TEUM2011",
  "message": "공통 가능한 시간대가 조회되었습니다.",
  "result": {
    "date": "2025-08-02",
    "availableTime": [
      {
        "start": "00:30",
        "end": "02:10"
      },
      {
        "start": "08:10",
        "end": "24:00"  // 24시 반환
      }
    ]
  }
}
```
- 8월 1일 23시 40분 ~ 8월 2일 0시 30분까지의 스케줄이 존재한다는 전제

yangjiae12 committed:
Feat: 종료 시각의 00:00을 LocalTime.MAX로 해석하여 시간 순서 검증

yangjiae12 committed:
Feat: TimeUtil에 종료 시간 24:00 변환 로직 추가

- 종료 시간이 LocalTime.MAX일 경우 "24:00" 문자열로 반환하는 formatEndTime() 메서드 추가
- convertEndTime() 메서드 주석 변경

yangjiae12 committed:
Refactor: TimeUtil 및 TeumConverter를 주입 방식으로 구조 개선

- TimeUtil을 TeumConverter에 주입하여 시간 포맷 유틸 메서드 사용 (formatEndTime 등)
- TeumConverter를 static 유틸에서 Spring Bean으로 전환
- TeumServiceImpl에서 TeumConverter를 명시적으로 주입받아 사용하도록 변경
- 전체 구조를 정적 유틸 호출 방식에서 DI 기반으로 리팩토링

yangjiae12 committed:
Feat: 24:00 시간 처리 유틸 추가

- convertEndTime(): 00:00 입력 시 LocalTime.MAX로 변환
- formatEndTime(): LocalTime.MAX → "24:00" 포맷
- parseAndFormatEndTime(): 변환 + 포맷 통합

yangjiae12 committed:
Fix: Schedule 생성 시 종료 시간 보정 및 가능한 시간대 계산 개선

- Schedule 생성 시 endTime이 LocalTime.MAX인 경우 다음 날 00:00으로 보정하여 정확한 Duration 계산 가능하도록 수정
- invertScheduledToAvailable()에서 중복 및 겹치는 시간대 정리
- 가능한 시간대 응답 시 종료 시간이 LocalTime.MAX일 경우 "24:00"으로 포맷되도록 처리

yangjiae12 committed:
Fix: 00:00 종료 시간 처리 및 정렬 오류로 인한 스케줄 누락 해결

- 스케줄의 종료 시간이 00:00이고 날짜가 다음 날일 경우, 사용자 의도에 따라 "24:00"으로 변환
- 병합 및 반전 로직에서 "24:00"은 LocalTime.MAX로 비교되도록 처리
- getAvailableTime 응답 내 availableTime 정렬 시 parseTimeForSort() 적용하여 "00:00"이 앞에 오도록 보장
- parse 용도별로 parseTimeForCompare / parseTimeForSort 메서드 분리하여 정렬/비교 충돌 방지

yangjiae12 committed:
Chore: 주석 추가 및 불필요한 로그 제거

- getAvailableTime(), merge/invertScheduledTimeSlots 등 주요 로직에 설명 주석 추가
- 디버깅용 로그 출력 제거 (정렬 확인용 등)

yangjiae12 committed:
Fix: 수면 시간 종료 구간 누락으로 인한 잘못된 availableTime 수정

- 수면 시간이 23시 이후 시작되고 자정 넘어 끝나는 경우, "23:59 ~ 24:00"이 available로 오판되는 문제 발생
- 종료 시간을 "24:00"으로 수정하여 LocalTime.MAX로 정확히 인식되도록 변경
- 반전 로직에서 올바르게 비어 있는 시간 계산됨

yangjiae12 committed:
Fix: 하루를 넘는 일반 일정이 잘못된 TimeSlot으로 처리되는 문제 수정

- fromSchedule()에서 endTime이 다음 날로 넘어가는 경우, 오늘 기준 24:00까지만 포함되도록 처리
- "23:00 ~ 01:00" 같은 일정이 "23:00 ~ 24:00"으로 변환되어 오늘의 busy 구간으로 반영됨
- 병합 및 반전 로직에서 오해 없이 정상 처리되도록 보정

yangjiae12 committed:
Feat: 일정 정보를 날짜 단위로 자르는 sliceScheduleToDate() 도입

- 하루를 넘는 일정이 있을 경우, 주어진 날짜 안에 해당하는 시간만 반영되도록 TimeSlot으로 분할
- getAvailableTime() 내부에서 기존 fromSchedule() 사용 대신 sliceScheduleToDate() 적용
- 병합 및 반전 시 정확한 시간 계산을 위해 날짜 경계 기준으로 일정 슬라이싱

yangjiae12 committed:
Fix: cross-day 일정이 availableTime 계산에 반영되지 않던 문제 해결

- 기존 일정 조회는 startTime 기준 date 일치 조건만 사용하여, 전날 시작해 자정 넘긴 일정이 누락되는 문제가 있었음
- endTime ≥ today 00:00 && startTime < today+1 00:00 조건으로 조회 범위 확장
- getAvailableTime()에서 하루를 넘는 일정도 정확하게 sliceScheduleToDate()와 함께 반영되도록 개선
- 결과적으로 00:00 이후 잔여 일정도 포함되어 availableTime 계산 정확도 향상

yangjiae12 committed:
Merge branch 'develop' into feature/114-teum-time-formatting

Jinseong01 commented on src/main/java/umc/teumteum/server/domain/teum/service/TeumServiceImpl.java:
```diff
@@ -299,26 +302,35 @@ public ScheduledTeumCancelResponseDto cancelScheduledTeum(Long scheduleId, Long
+                    .filter(schedule ->
+                            !schedule.getEndTime().toLocalDate().isBefore(date) &&
+                                    schedule.getStartTime().toLocalDate().isBefore(date.plusDays(1))
+                    )
```
findOverlappingSchedules()의 조건과 해당 filter의 조건이 비슷해보이는데 서로 다른 역할을 하나요??

yangjiae12 replied:
처음에 날짜를 걸치는 스케줄이 반영되지 않는 문제를 확인하고 우선 서비스 구현체에서 filter 조건을 추가해 해결해 보려고 했습니다 그런데 실제 문제는 쿼리에서 `date` 값을 기준으로 스케줄을 제한하고 있었던 것이고, 그걸 뒤늦게 인지하고 쿼리를 수정하면서 문제를 해결하게 되었습니다...

당시 오류를 해결하는 데 급급해서 이미 서비스 쪽에 들어간 유사한 필터는 그대로 남게되었고 지금 보니 결과적으로 비슷한 기능이 중복된 상태가 된 것 같습니다 ㅜㅜ

그리고 아까 회의에서는 이 filter로 구현한 것 생각하고 날짜가 이틀이 아니라 3, 4일에 걸치면 어떻게 구현할 수 있는가... 에 대해 얘기를 했던 건데 filter 제거하고 쿼리만 사용하면 이 문제도 해결되겠네요 ㅎㅎ...

앞으로는 오류 해결 후에 전체 흐름도 꼭 정리하도록 하겠습니다!! 코드 꼼꼼하게 봐주셔서 감사해요...🙇

Jinseong01 commented on src/main/java/umc/teumteum/server/domain/teum/service/TeumServiceImpl.java:
```diff
@@ -353,12 +365,19 @@ public AvailableTimeResponseDto getAvailableTime(User user, AvailableTimeRequest
+        // 병합 → 반전 → 병합
+        List<TimeSlot> mergedBusy = teumConverter.mergeScheduledTimeSlots(scheduledSlots);
+        List<TimeSlot> rawAvailable = teumConverter.invertScheduledToAvailable(mergedBusy);
+        List<TimeSlot> cleanAvailable = teumConverter.mergeScheduledTimeSlots(rawAvailable);
```
일정이 있는 시간들을 모으고 -> 반전을 통해 비어있는 시간을 구한다 로 이해했는데요!
마지막 병합은 혹시 왜 필요한 건지 알 수 있을까요??

yangjiae12 replied:
테스트를 하던 중에 `23:59` - `24:00`의 구간이 발생하는 걸 확인하고 예상치 못한 시간 오차로 인해 생기는 10분 이하의 시간대(앱의 전체 시간이 10분 단위니까...)를 정리해주는 로직을 넣으려고 우선 대충~... 최종 병합의 자리를 만들어 두고나서 다른 부분을 해결하면서 1분짜리 구간 발생 오류를 해결했는데 그러고나서 지우지 않았네요...

마지막 병합은 정리하도록 하겠습니다!! 리뷰 남겨주신 건 오전 중으로 수정하고 리뷰 요청 누를테니 그때 한번만 더 확인해주시면 감사하겠습니다!...

yangjiae12 committed:
Refactor: 쿼리와 중복되는 날짜 필터 및 불필요한 마지막 병합 제거

- getAvailableTime 내부에서 사용되던 날짜 범위 필터는 쿼리 조건(findOverlappingSchedules)과 중복되어 제거
  - 기존 filter는 이틀에 걸친 일정만 고려하지만, 쿼리는 그 이상 걸치는 일정도 정확히 반영 가능
- 반전된 availableTime에 수행하던 마지막 병합도 실제 기능에 기여하지 않아 함께 제거
  - 병합은 인접한 경우에만 작동하므로, 의도했던 10분 이하 시간 제거 목적엔 부적합

Jinseong01 reviewed (approved):
확인했습니다! 수고하셨습니다!
